app/Libraries/BeatmapsetDiscussionVoteService.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**
+ *    Copyright 2015-2017 ppy Pty. Ltd.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace App\Libraries;
+
+use App\Traits\Validatable;
+use DB;
+
+class BeatmapsetDiscussionVoteService
+{
+    use Validatable;
+
+    const VOTABLE_BEATMAPSET_STATES = ['graveyard', 'wip', 'pending'];
+
+    private $discussion;
+    private $params;
+
+    public function __construct($discussion, $params)
+    {
+        $this->discussion = $discussion;
+        $this->params = $params;
+    }
+
+    public function isValid()
+    {
+        $this->validationErrors()->reset();
+
+        if ($this->discussion->isLocked()) {
+            $this->validationErrors()->add('base', '.locked');
+        }
+
+        if (!in_array($this->discussion->beatmapset->status(), static::VOTABLE_BEATMAPSET_STATES, true)) {
```

Comment:
Apparently graveyard shouldn't be a votable state unless the voter is BN or above 🤷‍♂️